source/known-hook-events.ts

```diff
@@ -1,8 +1,81 @@
-export default [
+import {IncomingMessage} from 'http';
+import {Options, CancelableRequest} from './utils/types';
+import {HTTPError} from './errors';
+
+export type InitHook = (options: Options) => void;
+// TODO: The `Error` type should confirm to any possible extended error type that can be thrown. See https://github.com/sindresorhus/got#hooksbeforeerror
+export type BeforeErrorHook = (error: Error) => Error | void | Promise<Error> | Promise<void>;
+export type BeforeRequestHook = (options: Options) => void | Promise<void>
+export type BeforeRedirectHook = (options: Options) => void | Promise<void>
+export type BeforeRetryHook = (options: Options, error: HTTPError, retryCount: number) => void | Promise<void>
+export type AfterResponseHook = (response: IncomingMessage, retryWithMergedOptions: (options: Options) => IncomingMessage | CancelableRequest<IncomingMessage>) => IncomingMessage | CancelableRequest<IncomingMessage>
```

Comment:
I think it should be

```diff
-retryWithMergedOptions: (options: Options) => IncomingMessage | CancelableRequest<IncomingMessage>
+retryWithMergedOptions: (options: Options) => CancelableRequest<IncomingMessage> 
```

Reply:
I think this may be a confusion with the line as it is currently:

```ts
retryWithMergedOptions: (options: Options) => CancelableRequest<IncomingMessage>
```

I'll make a type for `RetryFunction` instead to make it more readable.